spi_nand_flash/priv_include/nand.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * SPDX-FileCopyrightText: 2022 mikkeldamsgaard project
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * SPDX-FileContributor: 2015-2024 Espressif Systems (Shanghai) CO LTD
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include "spi_nand_flash.h"
+#include "freertos/semphr.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define INVALID_PAGE 0xFFFF
+
+typedef enum {
+    STAT_ECC_OK = 0,
+    STAT_ECC_1_TO_3_BITS_CORRECTED = 1,
+    STAT_ECC_BITS_CORRECTED = STAT_ECC_1_TO_3_BITS_CORRECTED,
+    STAT_ECC_NOT_CORRECTED = 2,
+    STAT_ECC_4_TO_6_BITS_CORRECTED = 3,
+    STAT_ECC_MAX_BITS_CORRECTED = STAT_ECC_4_TO_6_BITS_CORRECTED,
+    STAT_ECC_7_8_BITS_CORRECTED = 5,
+    STAT_ECC_MAX
+} ecc_status_t;
+
+typedef struct {
+    uint8_t ecc_status_reg_len_in_bits;
+    uint8_t ecc_data_refresh_threshold;
+    ecc_status_t ecc_corrected_bits_status;
+} ecc_data_t;
+
+typedef struct {
+    uint8_t log2_page_size; //is power of 2, log2_page_size shift (1<<log2_page_size) is stored to page_size
+    uint8_t log2_ppb;  //is power of 2, log2_ppb shift ((1<<log2_ppb) * page_size) will be stored in block size
+    uint32_t block_size;
+    uint32_t page_size;
+    uint32_t num_blocks;
+    uint32_t read_page_delay_us;
+    uint32_t erase_block_delay_us;
+    uint32_t program_page_delay_us;
+    ecc_data_t ecc_data;
+} spi_nand_chip_t;
+
+typedef struct {
+    esp_err_t (*init)(spi_nand_flash_device_t *handle);
+    esp_err_t (*deinit)(spi_nand_flash_device_t *handle);
+    esp_err_t (*read)(spi_nand_flash_device_t *handle, uint8_t *buffer, sector_t sector_id);
+    esp_err_t (*write)(spi_nand_flash_device_t *handle, const uint8_t *buffer, sector_t sector_id);
+    esp_err_t (*trim)(spi_nand_flash_device_t *handle, sector_t sector_id);
+    esp_err_t (*sync)(spi_nand_flash_device_t *handle);
+    esp_err_t (*copy_sector)(spi_nand_flash_device_t *handle, sector_t src_sec, sector_t dst_sec);
+    esp_err_t (*get_capacity)(spi_nand_flash_device_t *handle, sector_t *number_of_sectors);
+} spi_nand_ops;
+
+struct spi_nand_flash_device_t {
+    uint8_t dev_index;
+    spi_nand_flash_config_t config;
+    spi_nand_chip_t chip;
+    const spi_nand_ops *ops;
+    uint8_t *work_buffer;
+    uint8_t *read_buffer;
+    SemaphoreHandle_t mutex;
+};
+
+esp_err_t register_nand_dev(spi_nand_flash_device_t *handle);
+esp_err_t unregister_nand_dev(spi_nand_flash_device_t *handle);
```

Comment:
Even though these are not public functions, let's namespace them to avoid link time conflicts. E.g. `nand_register_dev`, `nand_unregister_dev` (or whichever prefix you pick for the non-public functions in this component)